File: src/main/java/io/github/mdsimmo/bomberman/commands/CommandGroup.java
package io.github.mdsimmo.bomberman.commands;

import io.github.mdsimmo.bomberman.messaging.Message;
import io.github.mdsimmo.bomberman.messaging.Text;
import org.bukkit.command.CommandSender;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public abstract class CommandGroup extends Cmd {

    List<Cmd> children = new ArrayList<>();

    public CommandGroup(Cmd parent, Cmd ... children) {
        super(parent);
        addChildren(children);
    }

    /**
     * Adds some child commands. Should be called from the constructor of every implementation
     *
     * @param children the child commands
     */
    protected void addChildren(Cmd... children) {
        this.children.addAll(Arrays.asList(children));
    }

    @Override
    public List<String> options(CommandSender sender, List<String> args) {
        if (args.size() != 1)
            return null;
        List<String> options = new ArrayList<>();
        for (Cmd c : children) {
            options.add(c.name().toString());
        }
        return options;
    }

    @Override
    public Message extra() {
        return Text.COMMAND_GROUP_EXTRA.format();
    }

    @Override
    public Message example() {
        return context(Text.COMMAND_GROUP_EXAMPLE).format();
    }

    public Message usage() {
        return context(Text.COMMAND_GROUP_USAGE).format();
    }

    @Override
    public boolean run(CommandSender sender, List<String> args) {
        if (args.size() == 0) {
            help(sender);
            return true;
        } else {
            for (Cmd c : children) {
                if (c.name().toString()
                        .equalsIgnoreCase(args.get(0))) {
                    args.remove(0);
                    return c.execute(sender, args);
                }
            }
            context(Text.UNKNOWN_COMMAND)
                    .with("attempt", args.get(0))
                    .sendTo(sender);
            help(sender);
            return true;
        }
    }

    public Cmd getCommand(CommandSender sender, List<String> args) {
        if (args.size() == 0)
            return this;
        for (Cmd cmd : children) {
            if (cmd.name().toString().equals(args.get(0))
                    && cmd.isAllowedBy(sender)) {
                args.remove(0);
                if (cmd instanceof CommandGroup)
                    return ((CommandGroup) cmd).getCommand(sender, args);
                else
                    return cmd;
            }
        }
        return this;
    }

}

File: src/main/java/io/github/mdsimmo/bomberman/commands/CommandGroup.kt
package io.github.mdsimmo.bomberman.commands

import io.github.mdsimmo.bomberman.messaging.CollectionWrapper
import io.github.mdsimmo.bomberman.messaging.Message
import io.github.mdsimmo.bomberman.messaging.Text
import org.bukkit.command.CommandSender
import java.util.*

abstract class CommandGroup(parent: Cmd?) : Cmd(parent) {
    private val children: MutableList<Cmd> = ArrayList()

    /**
     * Adds some child commands. Should be called from the constructor of every implementation
     *
     * @param children the child commands
     */
    fun addChildren(vararg children: Cmd) {
        this.children.addAll(listOf(*children))
    }

    private fun child(sender: CommandSender, args: List<String>): Cmd? {
        return args.firstOrNull()?.let {arg ->
            children.filter { it.permission().isAllowedBy(sender) }
                    .firstOrNull { c -> c.name().toString().equals(arg, ignoreCase = true) }
        }
    }

    override fun options(sender: CommandSender, args: List<String>): List<String> {
        return when (val c = child(sender, args)) {
            null -> children
                    .filter { it.permission().isAllowedBy(sender) }
                    .map { it.name().toString() }
            else -> {
                c.options(sender, args.drop(1))
            }
        }
    }

    override fun flags(sender: CommandSender, args: List<String>, flags: Map<String, String>): Set<String> {
        return when (val c = child(sender, args)) {
            null -> emptySet()
            else -> c.flags(sender, args.drop(1), flags)
        }
    }

    override fun flagOptions(sender: CommandSender, flag: String, args: List<String>, flags: Map<String, String>): Set<String> {
        return when (val c = child(sender, args)) {
            null -> emptySet()
            else -> c.flagOptions(sender, flag, args.drop(1), flags)
        }
    }

    override fun help(sender: CommandSender, args: List<String>, flags: Map<String, String>) {
        when (val c = child(sender, args)) {
            null -> context(Text.COMMAND_GROUP_HELP).sendTo(sender)
            else -> c.help(sender, args.drop(1), flags)
        }
    }

    override fun extra(): Message {
        return Text.COMMAND_GROUP_EXTRA.format()
    }

    override fun example(): Message {
        return context(Text.COMMAND_GROUP_EXAMPLE).format()
    }

    override fun usage(): Message {
        return context(Text.COMMAND_GROUP_USAGE).format()
    }

    override fun run(sender: CommandSender, args: List<String>, flags: Map<String, String>): Boolean {
        return if (args.isEmpty()) {
            help(sender, args, flags)
            true
        } else {
            for (c in children) {
                if (c.name().toString().equals(args[0], ignoreCase = true)) {
                    c.execute(sender, args.drop(1), flags)
                    return true
                }
            }
            context(Text.UNKNOWN_COMMAND)
                    .with("attempt", args[0])
                    .sendTo(sender)
            help(sender, args, flags)
            true
        }
    }

    override fun format(args: List<Message>, elevated: Boolean): Message {
        return if (args.getOrNull(0).toString().equals("children", ignoreCase = true)) {
            CollectionWrapper(children).format(args.drop(1), elevated)
        } else {
            super.format(args, elevated)
        }
    }
}